enterprise/ha/src/main/java/org/neo4j/kernel/ha/MasterTransactionCommitProcess.java

```diff
@@ -19,46 +19,72 @@
  */
 package org.neo4j.kernel.ha;
 
+import java.io.IOException;
+
+import org.neo4j.helpers.collection.Visitor;
 import org.neo4j.kernel.api.exceptions.TransactionFailureException;
 import org.neo4j.kernel.ha.transaction.TransactionPropagator;
 import org.neo4j.kernel.impl.api.TransactionCommitProcess;
 import org.neo4j.kernel.impl.api.TransactionToApply;
+import org.neo4j.kernel.impl.locking.Locks;
+import org.neo4j.kernel.impl.transaction.command.Command.NodeCommand;
+import org.neo4j.kernel.impl.transaction.command.Command.PropertyCommand;
 import org.neo4j.kernel.impl.transaction.state.IntegrityValidator;
 import org.neo4j.kernel.impl.transaction.tracing.CommitEvent;
+import org.neo4j.storageengine.api.StorageCommand;
 import org.neo4j.storageengine.api.TransactionApplicationMode;
 
+import static org.neo4j.kernel.impl.api.index.NodePropertyCommandsExtractor.mayResultInIndexUpdates;
+import static org.neo4j.kernel.impl.locking.ResourceTypes.SCHEMA;
+import static org.neo4j.kernel.impl.locking.ResourceTypes.schemaResource;
+
 /**
  * Commit process on the master side in HA, where transactions either comes in from slaves committing,
  * or gets created and committed directly on the master.
  */
 public class MasterTransactionCommitProcess implements TransactionCommitProcess
 {
+    /**
+     * Detector of transactions coming in from slaves which should acquire the shared schema lock before
+     * being applied (and validated for application).
+     */
+    private static final Visitor<StorageCommand,IOException> REQUIRES_SHARED_SCHEMA_LOCK = command ->
+            command instanceof NodeCommand && mayResultInIndexUpdates( (NodeCommand) command ) ||
+            command instanceof PropertyCommand && mayResultInIndexUpdates( (PropertyCommand) command );
```

Comment:
On way, but we have not considered HA yet. Maybe when the functionality exists, we can use it here too. Not in scope for this review, but we cannot lock on individual indexes?